src/core/perf_event_array/v0_4_0.rs

```diff
@@ -0,0 +1,120 @@
+use crate::cpuonline;
+use crate::types::MutPointer;
+
+use bcc_sys::bccapi::*;
+use std::collections::HashMap;
+use std::hash::{Hash, Hasher};
+
+#[derive(Debug)]
+pub struct PerfEventArray {
+    name: String,
+    ev_type: u32,
+    ev_config: u32,
+    table_fd: i32,
+    cpu_fd: HashMap<usize, i32>,
+}
+
+impl PerfEventArray {
+    pub fn new(name: String, ev_type: u32, ev_config: u32, table_fd: i32) -> Self {
+        let hash_map: HashMap<usize, i32> = HashMap::new();
+        Self {
+            name,
+            ev_type,
+            ev_config,
+            table_fd,
+            cpu_fd: hash_map,
+        }
+    }
+
+    pub fn close_all_cpu(&mut self) {
+        self.cpu_fd.retain(|_, v| {
+            if *v < 0 {
+                unsafe { bpf_close_perf_event_fd(*v) };
+            }
+
+            false
+        });
+    }
+
+    pub fn close_on_cpu(&mut self, cpu: usize) {
+        let fd = self.cpu_fd.remove(&cpu);
+        if fd.is_some() {
+            unsafe { bpf_close_perf_event_fd(fd.unwrap()) };
+        }
+    }
+
+    pub fn open_all_cpu(&mut self) -> Result<(), String> {
+        let cpus = cpuonline::get();
+
+        if let Ok(cpus) = cpus {
+            for cpu in cpus {
+                let result = self.open_on_cpu(cpu);
+
+                if let Err(e) = result {
+                    self.close_all_cpu();
+                    return Err(e);
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    pub fn open_on_cpu(&mut self, cpu: usize) -> Result<(), String> {
+        if self.cpu_fd.get(&cpu).is_some() {
+            return Err("perf event is already open for cpu".to_string());
+        }
+
+        let fd =
+            unsafe { bpf_open_perf_event(self.ev_type, self.ev_config.into(), -1, cpu as i32) };
+
+        if fd < 0 {
+            return Err("failed to open perf on cpu".to_string());
+        }
+        let mut cpu_bytes: [u8; 8] = unsafe { std::mem::transmute(cpu.to_le()) };
```

Comment:
Wondering about the potential for endianness issues here. Thoughts?

Reply:
I was wondering the same thing. usize did not have `to_ne` so I added this as a placeholder. I just realized that there is a `to_ne_bytes()` function so I will try to see if I can change it to that.